### PR TITLE
Unit-tests for additional groupby aggregators

### PR DESCRIPTION
**Pull request by JovanVeljanoski**
Added unit-tests for

- standard deviation aggregator for numeric types;
- mode (most common value) for string/object type.

### CODE REVIEW

File: tests/groupby_test.py [hunk 1]
```diff
@@ -140,14 +140,34 @@ def test_groupby_count():
...
+    assert dfg.g.tolist() == [3.112474899497183, 1.118033988749895, 1.5]
+
+
 def test_groupby_count_string():
     g = np.array([0, 0, 0, 0, 1, 1, 1, 1, 2, 2])
     s = np.array(list(map(str, [0, 0, 0, 0, 1, 1, 1, 1, 2, 2])))
     df = vaex.from_arrays(g=g, s=s)
     groupby = df.groupby('s')
-    dfg = groupby.agg({'c': vaex.agg.count('s')})
+    dfg = groupby.agg({'m': vaex.agg.mode('s')})
```
Comment (maartenbreddels) [inline tests/groupby_test.py]: I guess this should still be count? (copy paste bug?)

File: tests/groupby_test.py [hunk 2]
```diff
@@ -140,14 +140,34 @@ def test_groupby_count():
...
+
+def test_groupby_mode_string():
+    g = np.array([0, 0, 0, 0, 1, 1, 1, 1, 2, 2])
+    s = np.array(list(map(str, [0, 0, 0, 0, 1, 1, 1, 1, 2, 2])))
+    df = vaex.from_arrays(g=g, s=s)
+    groupby = df.groupby('s')
+    dfg = groupby.agg({'m': vaex.agg.mode('s')})
     assert dfg.s.tolist() == ['0', '1', '2']
-    assert dfg.c.tolist() == [4, 4, 2]
+    assert dfg.m.tolist() == ['0', '1', '2']
```
Comment (maartenbreddels) [inline tests/groupby_test.py]: I'm not sure the test as it is is very solid, it does not have within 1 group a mix of different 's', a more thorough test would test a mix, where in 1 group the order is different from the other etc, this test is too easy to pass.

File: tests/groupby_test.py (hunk 1 above)
Comment (JovanVeljanoski) [reply tests/groupby_test.py]: Ah yes, indeed. Fixed in the upcoming commit.

File: tests/groupby_test.py (hunk 2 above)
Comment (JovanVeljanoski) [reply tests/groupby_test.py]: Indeed, the test can be made more proper. I guess my idea was just to present a use-case. Better test is coming in commit soon.

**Comment by maartenbreddels** [comment]
Note that mode is not implemented yet.

File: packages/vaex-core/src/superagg.cpp [hunk 3]
```diff
@@ -653,6 +653,51 @@ class AggSum : public AggBase<StorageType, StorageType, IndexType> {
...
+                        continue;
+                    this->grid_data[indices1d[j]] += pow(value, moment);
+                }
+            }
+        } else {
+            for(size_t j = 0; j < length; j++) {
+                StorageType value = this->data_ptr[offset + j];
+                if(FlipEndian)
+                    value = _to_native(value);
+                if(value == value) // nan check
```
Comment (manodeep) [inline packages/vaex-core/src/superagg.cpp]: Thank you again for creating and maintaining `vaex`. Hope this is helpful.

This `value == value` condition will be optimised away by most modern compilers when compiling with any optimisation flags, and therefore will not work as expected. One way to check for NANs (at least in `C`) is to add a `volatile` modifier to variable "value". 

I also notice that you are using `pow` for computing the moment, where `moment` is a `size_t`. Might be faster to compute the chained products for `moment = [1, 2, 3] -> [value, value*value, value*value*value]` and then default to `pow` for the general case. 

Thanks!

Comment (maartenbreddels) [reply packages/vaex-core/src/superagg.cpp]: Thank you!

value == value will indeed be optimized away with `--ffast-math`, but I've never seen it been optimized away without it (the unittest will trigger it). But I've never really dug into what is the fasted nan test.

I really like the idea for the moments, I think we should have a moments aggregator, that will do this (up to a particular moment order)! Would you mind opening an issue for this?

cheers,

Maarten

Comment (manodeep) [reply packages/vaex-core/src/superagg.cpp]: Interesting point about the NAN check. All of my reading implies that the use of `volatile` is necessary to ensure correct behaviour. May be that relates to `icc` being over-eager and enabling some degree of `ffast-math` by default. My strategy is always to call `isfinite(float)` and not worry about the run-time performance of the resulting call. (On my OSX laptop, `math.h` does contain two implementations for `isfinite` -> one for the "normal" case, and another for when `ffast-math` is enabled.)

I will open the issue regarding the moments. This is the `C` code I would write for that implementation (likely to be faster than an equivalent `switch` statement):

```C
const StorageType moment_values[] = {1, value, value*value, value*value*value};
const size_t max_moment = sizeof(moment_values)/sizeof(StorageType);
this->grid_data[indices1d[j]] += (moment < max_moment) ? moment_values[moment]: pow(value, moment);
```

Comment (manodeep) [reply packages/vaex-core/src/superagg.cpp]: I will note that my implementation above does not check for overflow/underflow. A better implementation would use an inter-mediate (larger-width) variable to hold the sum, and then cast down to the appropriate type at the end. Plus, depending on how the sums are evaluated, there might need to be a compensated algorithm (e.g, Kahan sum etc)

**Pull request merged**